### ISSUE TITLE
Buildbot: More builders, add benchmarks, test proxy.

### ISSUE THREAD

**Issue by dormando** [memcached/memcached#1024]
Continuing from #664 - I have new hardware to use for buildbots and automated benchmark testing. This issue is a worklog for setting this up. Expecting this to take at least a few weeks.
---
Current:
- Public interface: https://build.memcached.org/#/
- Small droplet running buildbot.
- Ancient dualcore intel NUC that runs builder VM's
- Pair of donated cloud metal instances for manual-ish benchmarking (now retired by the company)

New:
- Same small droplet, but upgraded to latest buildbot.
- 16 core 7950x + 64G RAM for VM's and mcshredder/mc-crusher bench testing.
- Many more builder VM's, as originally planned in #664 - test TLS, proxy, etc.
- Can add more hardware if necessary.

Part 1: Buildbot:
- [x] Upgrade buildbot to latest version.
- [x] Upgrade buildbot host to latest debian.
- [x] Migrate buildbot VM system to new hardware.
- [x] ~Replace buildbot VM's (password lost, OS's old)~ found the password. can keep using old VM's.
- [x] Add new build iterations with TLS + proxy on/off
- [x] ~Attempt to get kickoff-on-callback BB scheduler to work~ was already doing this; just cut wait time from 5m to 30s
- [x] Test `buildbot try` with new setup. For testing PR's. - already almost had this working! seems to work now.

Part 2: Benchmarking:
- [x] Add "Benchmark builders" which lock the entire machine as a resource so VM's and perf tests can't run at the same time. edit: got the locks added, no builders configured yet.
- [ ] Tune hardware settings (turbo, clock speeds, etc) until benchmarks run throughput and latency more repeatably.
- [x] Create scripts for sub-containers to start up with tests (mostly for proxy).
- [x] Add modes for limiting/breaking the network of sub-containers (bw limits, packet loss, latency, etc)
- [x] Run some manual benchmarks to get a feel for how to structure this with scripting.

Part 3: Automated regressions:
- [x] Improve mcshredder (probably track in separate issue). Have a bunch of things in mind for this.
- [ ] Create a separate repository for performance and functionality regression testing. (should allow 3rd party contributions)
- [ ] Look into auto-generating reports; ie throughput/latency per-release tag or per-commit.
- [x] Refactor proxy perl tests for dev efficiency; separate TODO (code cleanup, remove port usage, etc)
- [ ] Remove hardcoded backend ports from proxy perl tests (proxy already uses domain socket)

The final product should be infrastructure that can run tests submitted by anyone; and all test suites should be easy to clone and run in other environments - as is currently the case with perl tests.

List of ideas for builders:
- `make distclean`
- `make distcheck` - looks like this could be some work. Might write a smaller tarball validator script to start...
- something dockerish or vmish for pulling in a testing distro with newer compilers (debian testing? alpine?)
- `--disable-extstore` `--enable-seccomp` `--enable-tls` `--enable-proxy`; possibly one with a bunch of settings at once.
- valgrind? ASAN?
- big endian test? I did this once with qemu but have lost the setup already.
- newer/replaced OS builds for what I currently have. what replaces centos? Anything new worth testing?
- No way to legally do mac os x via VM I guess? don't want to spend $600 for a mac mini buildbot...

**Comment by dormando** [comment]
Ended up finding a bug in a fix I made last week because the test suite didn't run right last week :( Lame.

All the buildbot cruft is updated. Workers migrated to the new machine. Next step is to add a ton of builders since I can run 32-48 of them in parallel now.

**Comment by dormando** [comment]
Note: Before I add proxy tests to a bunch more builders, the proxy needs to be updated to allow unix sockets for backends.

In the meantime there is one slow builder which runs the proxy tests and `buildbot try` now works and runs on my full set of VM workers... so that's 90% of the way there and I can come back to adding/fixing builders after the next stage.

**Comment by dormando** [comment]
Stopping today with a bunch more of the buildbot stuff done. Turns out I had done most of it a few years ago and it just needed some tuning up. A few of the new builders will fail until later fixed; that's fine.

Got a few more upstream bugs fixed.

Continuing tomorrow with the container work. I have them starting, need to do internet and traffic shaping. Probably create a few and snapshot/clone them.

Seems like for performance tests I should only be using the containers on hosts which are intended to be broken (traffic shaped, firewalled, etc), otherwise the CPU overhead of the bridge is pretty high.

**Comment by dormando** [comment]
Got the container stuff partially figured out. People have so catastrophically over-compilcated containers. I blame all of you for this (yes, you).

Had a struggle in figuring out how to limit bandwidth in both directions, having to actually re-learn how tc works from scratch. Except I realized there's a virtual device on the host end, and I can just set two "outbound" qdisc's for rate limiting and that's completely fine.

At this point I mostly move on to part 3 for a while. It's time to iterate on mcshredder.

**Comment by dormando** [comment]
been under the weather; still chipping at this.

**Comment by dormando** [comment]
Did the unit test cleanup (fixed a bug!) finally today. Wasn't as bad as expected, should've done this first.

Delayed dealing with the rest of the hard coded ports since it's not a big deal. will get to it after I work on mcshredder for a while.

**Comment by dormando** [comment]
mcshredder's far enough along for me to come back to this...

next step is to figure out management of backend containers for tests and get some really basic stuff working. Then revisit it later with tuning to get high detail commit to commit performance reports.

For backend nodes I'm going to prebuild a few containers and set up some scripts ahead of time, then allow mcshredder to execute scripts which start/stop/adjust the containers. Then it should run through a series of tests for that particular configuration.

Once that POC works I'll see about creating a repository for sets of bench-based tests... should at least get a few dozen tests that run for a few minutes each with a proxy configuration. Probably going to end up needing to fix or add more to mcshredder as this goes along.

**Comment by dormando** [comment]
fml. keep getting sidetracked. still trying to find time to bootstrap this.

**Comment by dormando** [comment]
after chipping at it on and off for another month I finally have a suite-like thing that rolls through a ton of benchmark level stability tests. I'll be pushing the repository up soon.

I went back to manually running it as _running the  damn tests_ are more important than automation or reports or any other shit. It's already helped find three bugs and validate a PR that's been stuck since march.

I'm leaving this issue open until the suites are further along, publicly available, and perhaps some basic level of reporting or automation are done.

**Comment by dormando** [comment]
Still prioritizing fixing bugs that the suite is finding. It should be posted soon and tick off another TODO item here.

Nicer reporting would definitely be better but so far just logging the test run output and looking at it is netting benefits so I'm not in a rush..

**Comment by dormando** [comment]
been quietly chipping at this for a month. I have a usable test suite but unsure if/when I want to publish it publicly.

mcshredder has evolved a lot in the last month.

think I'll just close this issue as it's clear nobody's following the work log and this just wastes more of my time.
